cpp/src/strings/convert/convert_integers.cu

```diff
@@ -38,6 +37,147 @@
 
 namespace cudf {
 namespace strings {
+
+namespace detail {
+namespace {
+
+/**
+ * @brief This only checks if a string is a valid integer within the bounds of its storage type.
+ */
+template <typename IntegerType>
+struct string_to_integer_check_fn {
+  column_device_view const d_strings;
+  __device__ bool operator()(size_type idx) const
+  {
+    if (d_strings.is_null(idx)) { return false; }
+    auto const d_str = d_strings.element<string_view>(idx);
+    if (d_str.empty() || (d_str.data()[0] == '-' && std::is_unsigned<IntegerType>::value)) {
+      return false;
+    }
+
+    auto iter = d_str.data() + static_cast<int>((d_str.data()[0] == '-' || d_str.data()[0] == '+'));
+    auto const iter_end = d_str.data() + d_str.size_bytes();
+    if (iter == iter_end) { return false; }
+
+    auto const sign = d_str.data()[0] == '-' ? IntegerType{-1} : IntegerType{1};
+    auto const bound_val =
+      sign > 0 ? std::numeric_limits<IntegerType>::max() : std::numeric_limits<IntegerType>::min();
+
+    IntegerType value = 0;      // parse the string to integer and check for overflow along the way
+    while (iter != iter_end) {  // check all bytes for valid characters
+      auto const chr = *iter++;
+      // Check for valid character
+      if (chr < '0' || chr > '9') { return false; }
+
+      // Check for underflow and overflow:
+      auto const digit       = static_cast<IntegerType>(chr - '0');
+      auto const bound_check = (bound_val - sign * digit) / IntegerType{10} * sign;
+      if (value > bound_check) return false;
+      value = value * IntegerType{10} + digit;
+    }
+
+    return true;
+  }
+};
+
+/**
+ * @brief The dispatch functions for checking if strings are valid integers.
+ */
+struct dispatch_is_integer_fn {
+  template <typename T, std::enable_if_t<std::is_integral<T>::value>* = nullptr>
+  std::unique_ptr<column> operator()(strings_column_view const& strings,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr) const
+  {
+    auto const d_column = column_device_view::create(strings.parent(), stream);
+    auto results        = make_numeric_column(data_type{type_id::BOOL8},
+                                       strings.size(),
+                                       cudf::detail::copy_bitmask(strings.parent(), stream, mr),
+                                       strings.null_count(),
+                                       stream,
+                                       mr);
+
+    thrust::transform(rmm::exec_policy(stream),
+                      thrust::make_counting_iterator<size_type>(0),
+                      thrust::make_counting_iterator<size_type>(strings.size()),
```

Comment:
Instead of using counting iterators, using the pair iterator would be more natural and better express intent.